lib/codecs/src/decoding/format/avro.rs

```diff
@@ -0,0 +1,356 @@
+        Self {
+            avro_options: AvroDeserializerOptions {
+                schema,
+                strip_schema_id_prefix,
+            },
+        }
+    }
+
+    /// Build the `AvroDeserializer` from this configuration.
+    pub fn build(&self) -> AvroDeserializer {
+        let schema = apache_avro::Schema::parse_str(&self.avro_options.schema)
+            .map_err(|error| format!("Failed building Avro serializer: {}", error))
+            .unwrap();
+        AvroDeserializer {
+            schema,
+            strip_schema_id_prefix: self.avro_options.strip_schema_id_prefix,
+        }
+    }
+
+    /// The data type of events that are accepted by `AvroDeserializer`.
+    pub fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_definition(&self, log_namespace: LogNamespace) -> schema::Definition {
+        match log_namespace {
+            LogNamespace::Legacy => {
+                let mut definition = schema::Definition::empty_legacy_namespace()
+                    .unknown_fields(vrl::value::Kind::any());
+
+                if let Some(timestamp_key) = log_schema().timestamp_key() {
+                    definition = definition.try_with_field(
+                        timestamp_key,
+                        vrl::value::Kind::any().or_timestamp(),
+                        Some("timestamp"),
+                    );
+                }
+                definition
+            }
+            LogNamespace::Vector => schema::Definition::new_with_default_metadata(
+                vrl::value::Kind::any(),
+                [log_namespace],
+            ),
+        }
+    }
+}
+
+impl From<&AvroDeserializerOptions> for AvroSerializerOptions {
+    fn from(value: &AvroDeserializerOptions) -> Self {
+        Self {
+            schema: value.schema.clone(),
+        }
+    }
+}
+/// Apache Avro serializer options.
+#[configurable_component]
+#[derive(Clone, Debug)]
+pub struct AvroDeserializerOptions {
+    /// The Avro schema definition.
+    /// Please note that the following [`apache_avro::types::Value`] variants are currently *not* supported:
+    /// * `Date`
+    /// * `Decimal`
+    /// * `Duration`
+    /// * `Fixed`
+    /// * `LocalTimestampMicros`
+    /// * `LocalTimestampMillis`
+    /// * `TimeMillis`
+    /// * `TimestampMicros`
+    /// * `TimestampMillis`
+    ///
+    /// This seems to be an issue when converting Vrl back into avro.
+    /// Example:
+    /// timestamp_micros and timestamp_millis are longs in Avro but converted to timestamps for Vrl,
+    /// so when trying to change them back to Avro they fail with unexpected type.
+    /// This should be fixable by extending encode in `lib/codecs/src/encoding/format/avro.rs`
+    /// to know the avro schema at serialization time.
+    /// [See comment here](https://github.com/vectordotdev/vector/pull/19342#discussion_r1424215373)
```

Comment:
```suggestion
```

I would move the rest of the details under the relevant match block in `try_from`:
```
    /// Example:
    /// timestamp_micros and timestamp_millis are longs in Avro but converted to timestamps for Vrl,
    /// so when trying to change them back to Avro they fail with unexpected type.
    /// This should be fixable by extending encode in `lib/codecs/src/encoding/format/avro.rs`
    /// to know the avro schema at serialization time.
    /// [See comment here](https://github.com/vectordotdev/vector/pull/19342#discussion_r1424215373)
```

Reply:
Hmm, was just thinking that we could basically set them as i64 instead of parsing them as vrl timestamps.

Then leave it up to the user to convert in to timestamps using vrl.

Just changed it and the test parse.

```
        AvroValue::TimestampMicros(ts_micros) => Ok(VrlValue::from(ts_micros)),
        AvroValue::TimestampMillis(ts_millis) => Ok(VrlValue::from(ts_millis)),
```

Reply:
Hmm we could also do the same for date, it just expects users to know how their avro works, but think that is a fair assumption.

Reply:
[I added support for timestamps and local timestamps](https://github.com/vectordotdev/vector/pull/19342/commits/a669d7373a3bd404200c73922c813eac2cde3b54)

If you prefer to have it as timestamps in vrl i can revert this commit

Reply:
What does the encoder do? Is the roundtrip lossless?

Reply:
The avro libs own avro to json conversion maps to number in json
```
            Value::TimestampMillis(t) => Ok(Self::Number(t.into())),
            Value::TimestampMicros(t) => Ok(Self::Number(t.into())),
            Value::LocalTimestampMillis(t) => Ok(Self::Number(t.into())),
            Value::LocalTimestampMicros(t) => Ok(Self::Number(t.into())),
```

Have added roundtrip test for local timestamp and timestamp kept the original timestamp one, both pass.